src/Analyzers/Razor.Diagnostics.Analyzers/IntermediateTokenAnalyzer.cs

```diff
@@ -0,0 +1,127 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT license. See License.txt in the project root for license information.
+
+using System.Collections.Immutable;
+using System.IO;
+using System.Linq;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Operations;
+
+namespace Razor.Diagnostics.Analyzers;
+
+[DiagnosticAnalyzer(LanguageNames.CSharp)]
+public class IntermediateTokenAnalyzer : DiagnosticAnalyzer
+{
+    private const string Title = "IntermediateToken with Kind CSharp";
+    private const string MessageFormat = "Avoid directly creating an IntermediateToken with Kind 'CSharp'. Instead, lower to an appropriate intermediate node and emit the C# in the output writer.";
+    private const string BaselineFileName = "IntermediateTokenBaseline.txt";
+
+    internal static readonly DiagnosticDescriptor Rule = new(
+        DiagnosticIds.RawIntermediateTokenCreation,
+        Title,
+        MessageFormat,
+        DiagnosticCategory.Usage,
+        DiagnosticSeverity.Warning,
+        isEnabledByDefault: true,
+        helpLinkUri: "https://github.com/dotnet/razor/issues/11858");
+
+    public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => [Rule];
+
+    public override void Initialize(AnalysisContext context)
+    {
+        context.EnableConcurrentExecution();
+        context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.Analyze | GeneratedCodeAnalysisFlags.ReportDiagnostics);
+
+        context.RegisterOperationAction(AnalyzeObjectCreation, OperationKind.ObjectCreation);
+        context.RegisterOperationAction(AnalyzeMethodInvocation, OperationKind.Invocation);
+    }
+
+    private void AnalyzeObjectCreation(OperationAnalysisContext context)
+    {
+        var objectCreation = (IObjectCreationOperation)context.Operation;
+
+        // Check if the type is IntermediateToken
+        if (objectCreation.Type?.Name != "IntermediateToken")
+        {
+            return;
+        }
+
+        // Resolve the TokenKind.CSharp enum value
+        var tokenKindType = context.Compilation.GetTypeByMetadataName("Microsoft.AspNetCore.Razor.Language.Intermediate.TokenKind");
+        if (tokenKindType == null)
+        {
+            return; // TokenKind enum not found
+        }
+
+        var csharpEnumMember = tokenKindType.GetMembers("CSharp").FirstOrDefault() as IFieldSymbol;
+        if (csharpEnumMember == null)
+        {
+            return; // TokenKind.CSharp enum member not found
+        }
+
+        // Check if the Kind property is set to TokenKind.CSharp via an initializer
+        var kindInitializer = objectCreation.Initializer?.Initializers
+            .OfType<ISimpleAssignmentOperation>()
+            .FirstOrDefault(init =>
+                init.Target is IPropertyReferenceOperation property &&
+                property.Property.Name == "Kind" &&
+                init.Value.ConstantValue.HasValue &&
```

Comment:
It's safest to have this as `init.Value.WalkDownConversion()` (you can borrow it from here https://github.com/dotnet/roslyn-analyzers/blob/a5e9b9b92f25491429401a3e43914ad1ac161357/src/Utilities/Compiler/Extensions/IOperationExtensions.cs#L701-L714)

In cases of conversions, IIRC, you can only see the constant value, if any, from the conversion operation operand.